File: blind 75 leetcode/Trees/invert_binary_tree.py
'''
Given the root of a binary tree, invert the tree, and return its root.
https://leetcode.com/problems/invert-binary-tree/
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if (root):
            ## swap the values of the left and right node
            tmp = root.left
            root.left = root.right
            root.right = tmp
            ## recursively run down the left side 
            self.invertTree(root.left)
            ## recursively run down the right side
            self.invertTree(root.right)
        ## once bottom of the tree is reached, return the root since all the nodes have been swapped at this point
        return root

File: node_js_ds_and_algos_implementations/trees/invert_binary_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function(root) {
    if (root){
        // in lieu of python's swap syntax:
        /*
        we set a temp variable for root.left's value
        we set a temp2 variable for root.right's value
        
        Set root.right's value now to be the temp variable,
        which was root.left's value
        Set root.left's value now to be the temp2 variable,
        which was root.right's value
        */
        tmp = root.left
        tmp2 = root.right
        root.right = tmp
        root.left = tmp2
        invertTree(root.left)
        invertTree(root.right)
    }
    return root
};